### ISSUE TITLE
[QUESTION] P2P Network topology

### ISSUE THREAD

**Issue by folkvir** [amark/gun#587]
Hi,

First, thank you for your work. 👍 

**Question 1:** Is the current network topology a mesh network? 
**If yes:** this kind of topology is not churn resistant and could break your network and create network partitions. In this case you need a way to reconstruct a proper network topology. But even with this "magic" method you have to rely on an external user (eg: a user well-connected or a signaling server.)
**If not**: What is the the topology used?

**Question 2:** Is it possible to implement another network topology with the current GUN architecture? (which seems a bit hard to understand.)

Can I suggest you to read [this paper](https://link.springer.com/article/10.1007/s11280-017-0478-5)([pdf](https://link.springer.com/content/pdf/10.1007%2Fs11280-017-0478-5.pdf))? This is the adaptive version of [Cyclon](https://link.springer.com/article/10.1007/s10922-005-4441-x)([pdf](https://link.springer.com/content/pdf/10.1007%2Fs10922-005-4441-x.pdf)) called Spray.
Both of them are random peer sampling networks. But Cyclon relies on a fixed partial view of the network, whereas Spray is an adaptive version of Cyclon where the size of the partial view grows or decrease with the network size and is always bounded to log(N) (where N is the size of the network).

If creating a new and more resistant network is possible it could be very nice for you. Because your GUN system could be more resistant to churn. Also, if building a Spray network is possible you could also build any kind of topology based on any criteria.

**Comment by folkvir** [comment]
After reading more documentation, my questions are out of scope because Gun is not using a pure P2P network. It relies on a server to forward messages. So optimizations making the network more resistant have to just rely on better server performance (eg autoscaling if 1M peers are connected to your DB).  In fact, as I understand from your documentation, it is a design choice. But on my mind, this choice will not scale if the number of peers drastically grows.

**Comment by folkvir** [comment]
@amark Hum, a superpeer network could work. I agree. But it is not necessary. I am working in this field and I am able to create unstructured/structured networks of web browsers easily using WebRTC. There is no pure WebRTC constraints that restrict the usage of a distributed and decentralized database. Only the fact that the database has to work in the web browser and that routing algorithms are missing but even with theses constraints you can deal with this by implementing supports for webrtc/browser applications.

Also, in a WebRTC networks of browsers, a signaling server just acts during the first connection. Other connections created inside the network do not rely anymore on this signaling server. It means that only one exchange of offers is needed to join the network of browsers. Then, as usual, just sends messages to another peer in the network just by relying on the connection between this two peers, or forward/broadcast the message if you want to reach the whole network.

I agree that GUN is P2P by nature. But why decentralized? If the GUN's server is down, are you able to send messages/data to other peers? if not, so your architecture is completely centralized, there is only one point of failure in your system.

**Comment by amark** [comment]
@folkvir I really really really want WebRTC to be reliable. Even if what you say winds up working perfectly... there is still the problem that WebRTC connections aren't persistent, so a single browser refresh would kill all connections and they'd have to be bootstrapped again unfortunately. Do you have a workaround to this?

We're able to have N number of relay peers (and ANYBODY can run one), which = NOT centralized. Obviously, I want WebRTC... but as long as WebRTC has to bootstrap through STUN/TURN on every reload... even with WebRTC you'd still need N number of bootstrapping peers.

**Comment by folkvir** [comment]
@amark 
**tl;dr** My long monologue contains informations about how I build an unstructured network and how a signaling server acts during the first connection. Also, I present a way to use this network to make a simple decentralized and distributed database. I am also talking about consistency criteria, so if you want the whole story, read me. ;) And finally I just ask 2 questions: 
- is "ANYBODY" a web browser?
- is it possible with the current architecture to build another bootstrapping method? I mean, are we able to manage connections after the connection was created to the superpeer?

--------------
Bootstrapping only one peer is not the problem. But bootstrapping again a superpeer because this peer reloaded its page is the problem yes. But you cannot rely on the user intent as the user can do anything at anytime (refresh its page when he wants).
So finding a solution around this is only possible if Service Workers are able to maintain WebRTC connections or a single Extension (that currently works) or a third party process running in background. (for example an electron app). But, to my mind, these solutions are not "Web". 

Question: is "ANYBODY" a web browser?

I think the base model of your system can be enhanced by just relying on a P2P-WebRTC network with an unstructured network (eg: a random peer sampling protocol as presented in my first message). But in this case you have to completely rebuild your architecture for sending messages/operations. And retrieving data over this network remains challenging in a way that you have to broadcast operations and retrieve not received operations in order to converge to a state that is common to each peer. 

A simple example with peer A, B and C:
* A engages its connection on the network "http://example.com/room=1", here, example.com is the signaling server address. and the room is a room dedicated for the network (A, B, C)
* A sends offers to the signaling server, this server detects that A is alone in the room "1", A receive a message from the server saying that he is connected.
* B begins its connection now and want to join the room "1", B sends its offers, the server (S) sends offers to A which accepts and resend its accepted offers to S, S sends back these offers to B. B is now connected to A through a WebRTC connection. 
* C wants to join the same network, C sends its offers to S, S choose a random peer to exchange offers.  C is now connected to the network composed of A, B and C.

You have now a P2P network composed of A connected to B, B connected to C, C connected to A or C connected to B (it depends on the server choice).  It means that if A create/delete/update a data, operations have to be broadcasted to the network. So if C want to receive the data, B have to route the operation. But this is not the only one problem, you have to order operations such as Update in order to define what kind of consistency your system will support. One simple criteria you can implement is the eventual consistency or the causal consistency which implies update consistency.(if you are interested in consistency criteria, [read this book](https://books.google.fr/books?hl=fr&lr=&id=8ksgDgAAQBAJ&oi=fnd&pg=PP1&dq=mathieu+perrin+distributed+systems+concurrency+and+consistency&ots=D-B_aJw4Wr&sig=XBmylSdZybiHwX_45Yvlqa5-4N4#v=onepage&q=mathieu%20perrin%20distributed%20systems%20concurrency%20and%20consistency&f=false) [1])

Now you can connect any peer to another, creating a network of browsers. You can rely on a protocol which will manage these connections. This protocol will create, delete, manage connections without relying anymore on the signaling server. It means that your network, or the db implemented on this network will be completely decentralized. And in this case you do not have any superpeer. If you implement Cyclon[2] or Spray[3], you will have a random graph and more, resilient to churn. But as I said, you have to "rethink" your current system to make it works. 

Moreover, and I agree, for any connection you have to rely on STUN/TURN servers for bootstrapping a connection. But this is mandatory if you want to create a connection between to peers. So you have to deal with. An interesting update to this system could be to remove this mechanism. But now how to discover an my public IP address? Perhaps relying on an existing connection that already knows me...My thought needs to be deeper, perhaps the solution is this way.  

Moreover (again), if your system is "offline first", you are able to store operations without a connection, do your stuff and then you can send operations when the connection suddenly reappear. And get other missing operations from other peers. But in this specific case, you will (and I am sure you will) receive conflicted operations. But even in presence of conflicts, the consistency criteria will resolve them (with your specific conflict resolution policy of course). 

--------------
[1] PERRIN, Matthieu. Distributed Systems: Concurrency and Consistency. Elsevier, 2017.
[2] VOULGARIS, Spyros, GAVIDIA, Daniela, et VAN STEEN, Maarten. Cyclon: Inexpensive membership management for unstructured p2p overlays. Journal of Network and systems Management, 2005, vol. 13, no 2, p. 197-217.
[3] NÉDELEC, Brice, TANKE, Julian, FREY, Davide, et al. An adaptive peer-sampling protocol for building networks of browsers. World Wide Web, 2018, vol. 21, no 3, p. 629-661.

**Comment by folkvir** [comment]
@amark 
**I would like to know if it is possible to change the way you send/propagate your operations?** 
I would like to **test if a fully WebRTC support is possible**. 
Example: 
- Connect a peer to an existing network of browsers (already done),
- Use Gun.js as database, **but I need to get the class/file which send data** (for now to the server)
- When I have to send data, rely on the existing network to broadcast operations (already done)

Tell me if I am wrong, is it sufficient to change the mesh.js file to a WebRTC adapter?

**Comment by amark** [comment]
@folkvir apologies I was gone last week traveling and this week at a conference...

In short, answer is yes, communication / transport adapters are made as modules to they can be swapped with different systems. See https://github.com/amark/gun/blob/master/src/adapters/websocket.js and similar. 

Specifically, DAM (daisy chain ad hoc mesh network) is an abstraction for mesh network topology and provides some pre-built interfaces that you can plug into - unfortunately I have documented it yet.

Give me some time and I can help explain but will need to e after next week. @fuchidahiro should be able to help explain some of the I/O events that transport adapters get called with. Basically have WebRTC messages call in on GUN and listen to out and relay over WebRTC. I know WebRTC is non trivial to manage connections though, so happy to answer Qs later to help.

**Comment by fuchidahiro** [comment]
Hi @folkvir,

Let me know if you need any help on this. (I am available at our gitter chat too, if that is better for you). In general, you can set up GUN.on('in', callback) to handle incoming items. But you can call gun.on('in', msg) with the message you get from another peer via your webrtc connection. On the same note you can setup GUN.on('out', callback) to handle anything that would go out and send it to other peers via your webRTC connection.

Example code (for flow not for code accuracy):
```javascript
var GUN = require('gun')

GUN.on('in', (data)=>{console.log(data)}) //log anything that is coming in
GUN.on('out', callback) // handle outgoing items to send to peers

//pseudo callback 
function callback(msg) {
var array = RTCPeerConnections;
for(connection in array) {
 connection.send(msg)
}
};

var gun = GUN({webRTC: true}) //not a real option just putting for clarity

webRTC.on('data', intoGun);

//pseudofunction intoGun

function intoGun(msg){
gun.on('in', msg); //when doing gun in minor letters, you are referring to the actual gun instance used
}
```

Let me know if this works for you, it's super high level of course and I don't profess to know webRTC. (Just brief glance so far)

**Comment by fuchidahiro** [comment]
@folkvir Your welcome, always glad to help.
It would be fun to do a demo of your code where two laptops connect via the signalling server. Browsers are showing a chat with the two guys talking and then the signalling server dies and they keep writing truly P2P.  Are you saying this would be working at this point in time? Have you tried to do so in one of your tests?

**Comment by folkvir** [comment]
@fuchidahiro Oh yes, it will work I am absolutely sure. In my scenario (example) there is no Gun server and it works. I just updated my gist with your Chat example (https://gist.github.com/folkvir/d92c29c7b8cc5cfcd0228ae346028f41).
* Just download the gist and `npm install && npm start` in the folder and open the gun.html file in your browser (Chrome not firefox because there is no stun/turn added in the example).
* Open a second (private) tab and try to play. 
* Now kill the server and try to send a message. 
* It continues to work.

Of course WebRTC connection are not persistent, so you will have to recreate a connection each time you refresh the page. And if the signaling server is down you cant join the network anymore. 

But for now you have a pure P2P Gun App working as example.

**Comment by fuchidahiro** [comment]
@folkvir Awesome!! Tres bien! I saw you have another repo, that will create a peer that can act as the signalling server once in the network. If you have at least 3 peers, you can keep reconnecting after the time out. I am guessing, webRTC connections can't be kept alive? (like the old idea of long-polling or sending data every once in a while to keep it from dying

**Comment by folkvir** [comment]
* [neighborhood-wrtc](https://github.com/ran3d/neighborhood-wrtc) is a repo for creating a peer with N connection to other peers. It uses simple-peer as WebRTC wrapper but you can change to another system if you want to migrate to an IP system, or bluetooth system. Just respect the API and it will work.

>If you have at least 3 peers, you can keep reconnecting after the time out.

* [n2n-overlay-wrtc](https://github.com/ran3d/n2n-overlay-wrtc) is a repo for managing these connections, you can create/disconnect or even acts as a bridge for connecting neighbours between them (eg: A->B, A->C), you are able to do A.bridge(B, C) and it will create B->C without using the signaling server. That's why I said in all my previous messages we just rely on a server for the very first connection. Otherwise there is no server at all. Except for STUN/TURN server when a connection is opened. But this is mandatory cause WebRTC protocols work like this. 

* [spray-wrtc](https://github.com/ran3d/spray-wrtc) is a protocol for managing these N2N peers. A spray peer has a neighbour view of a*log(N)+b. This is very cool because if the network grows, the view also grows, if the network size decrease, the size of the view decrease. Connections are renewed periodically

In short, you can create a lot of networks using n2n-overlay-wrtc pretty easily.

[foglet-core](https://github.com/ran3d/foglet-core) is a wrapper around networks created with the N2N repo. We provide easy primitives around this such as unicast (send message from a peer to a neighbour), causal broadcast (broadcast the message to all peers in the network with an anti-entropy mechanism if necessary), unicast stream and broadcast stream. You can create different channel easily. We also provide (and will provide) other networks (cyclon) and more sooner. You also can add overlay networks to create different communication layer. For example, our library is already used in research prototypes. 

>I am guessing, webRTC connections can't be kept alive? (like the old idea of long-polling or sending data every once in a while to keep it from dying

The answer is no. But I think there is a thread started by **feross** years ago or months ago for doing this. The idea is to enable a way to transfer a connection. This will be usefull for Service Worker. And it could enable a way to create some Content Delivery Networks using WebRTC connection for example. But also for more things. 
There is also a reason why there is no persistence across connections because, for each tab there is a port allocated. So if you refresh the page, the port will be different. So the peer connected to you will never receive any more message from you. You have to re create the connection to send more messages.

**Comment by fuchidahiro** [comment]
@folkvir Thanks for the explanation! It's some really cool stuff.

**Comment by amark** [comment]
@folkvir , @Dletta has brought this thread back to my attention... I need to re-read through it (I'm terribly slow though).

I also see you popping up on the other GitHub issues with some great comments, @folkvir you are working with the SOLID people right? Most of my interactions with people are via the https://gitter.im/amark/gun when I have time, so I definitely want to make sure you don't feel left out because I'm slow/delayed replying to Issues. Apologies!

**Comment by amark** [comment]
@folkvir wow! Just got to read and see the code with the chat example on saving data to GUN and being networked by your systems with WebRTC underneath! Wow!!!! Amazing!!!!

Your Spray/N2N logic looks very valuable (I was going to need to implement some sort of intelligent WebRTC sharding, using this instead will be very nice!)

DAM (part of GUN's stack already) does very efficient P2P robust messaging (what you call causal), I'm not sure its performance compared to your anti-entropy, but I was able to get it down to `P-1` where P is number of peers in the network (assuming uniform distribution of peers in a mesh network), compared to "brute force" which is `C*2+1` (if I recall correctly), where C is the number of connections.

Without actual sharding the WebRTC connections, like what spray does, there isn't much gain - so spray looks great. Two questions:

1. Would you be open to more tightly integrating the two?
2. How modular is the math/logic around spray? I love Feross, but simple-peer AND/OR webrtc has been spotty. Certainly, this could be just WebRTC's fault, not simple-peer, IDK (and part of the reason why although GUN is decentralized we've still relied on websockets these past years), so requiring a bunch of extra dependencies is not ideal.

Thoughts?
